lms/djangoapps/courseware/masquerade.py

```diff
@@ -116,3 +179,83 @@ def get_masquerading_group_info(user, course_key):
     if not course_masquerade:
         return None, None
     return course_masquerade.group_id, course_masquerade.user_partition_id
+
+
+# Sentinel object to mark deleted objects in the session cache
+_deleted_sentinel = object()
+
+
+class MasqueradingKeyValueStore(KeyValueStore):
+    """
+    A `KeyValueStore` to avoid affecting the user state when masquerading.
+
+    This `KeyValueStore` wraps an underlying `KeyValueStore`.  Reads are forwarded to the underlying
+    store, but writes go to a Django session (or other dictionary-like object).
+    """
+    def __init__(self, kvs, session):
+        """
+        Arguments:
+          kvs: The KeyValueStore to wrap.
+          session: The Django session used to store temporary data in.
+        """
+        self.kvs = kvs
+        self.session = session
+        self.session_data = session.setdefault(MASQUERADE_DATA_KEY, {})
+
+    def _serialize_key(self, key):
+        """
+        Convert the key of Type KeyValueStore.Key to a string.
+
+        Keys are not JSON-serializable, so we can't use them as keys for the Django session.
+        The implementation is taken from cms/djangoapps/contentstore/views/session_kv_store.py.
+        """
```

Comment:
@cpennington I don't think so.  `session_kv_store.py` both reads from and writes to the session, whereas this key-value store uses the session as a layer on top of some underlying key-value store, which is quite different.  Another difference is that the `session_kv_store.py` implementation dumps the data directly into the session, resulting in a lot of entries with weird keys directly in the session dictionary, which is (a) weird and (b) impossible to clean up.

Even if we fix this weirdness, there would be only a single method (`set()`) with identical implementation, and I don't think that trivial one-liner warrants moving a module around.